File: kafka-clients/python/kafka_client.py
from confluent_kafka import Producer, Consumer
import os
import json
from typing import Dict, Any, Optional

class KafkaClient:
    def __init__(self, config_path: str = "kafka-clients/python/client.properties"):
        self.config = self._read_config(config_path)
    
    def _read_config(self, config_path: str) -> Dict[str, str]:
        """Reads the client configuration from client.properties"""
        config = {}
        try:
            with open(config_path) as fh:
                for line in fh:
                    line = line.strip()
                    if len(line) != 0 and line[0] != "#":
                        parameter, value = line.strip().split('=', 1)
                        config[parameter] = value.strip()
        except FileNotFoundError:
            print(f"Config file {config_path} not found. Using environment variables.")
            # Fallback to environment variables
            config = {
                'bootstrap.servers': os.getenv('KAFKA_BOOTSTRAP_SERVERS', 'localhost:9092'),
                'security.protocol': os.getenv('KAFKA_SECURITY_PROTOCOL', 'PLAINTEXT'),
                'sasl.mechanisms': os.getenv('KAFKA_SASL_MECHANISMS', ''),
                'sasl.username': os.getenv('KAFKA_SASL_USERNAME', ''),
                'sasl.password': os.getenv('KAFKA_SASL_PASSWORD', ''),
                'client.id': os.getenv('KAFKA_CLIENT_ID', 'easyflow-python-client')
            }
            # Remove empty values
            config = {k: v for k, v in config.items() if v}
        return config

    def produce_message(self, topic: str, message: Dict[str, Any], key: Optional[str] = None) -> bool:
        """Produces a message to the specified topic"""
        try:
            producer = Producer(self.config)
            
            # Convert message to JSON string
            message_value = json.dumps(message) if isinstance(message, dict) else str(message)
            
            # Produce the message
            producer.produce(
                topic, 
                key=key, 
                value=message_value,
                callback=self._delivery_callback
            )
            
            # Wait for message to be delivered
            producer.flush()
            return True
            
        except Exception as e:
            print(f"Error producing message: {e}")
            return False
    
    def _delivery_callback(self, err, msg):
        """Callback for message delivery confirmation"""
        if err is not None:
            print(f"Message delivery failed: {err}")
        else:
            print(f"Message delivered to {msg.topic()} [{msg.partition()}] at offset {msg.offset()}")

    def consume_messages(self, topic: str, group_id: str = "easyflow-python-group", auto_offset_reset: str = "earliest"):
        """Consumes messages from the specified topic"""
        consumer_config = self.config.copy()
        consumer_config.update({
            "group.id": group_id,
            "auto.offset.reset": auto_offset_reset
        })
        
        consumer = Consumer(consumer_config)
        consumer.subscribe([topic])
        
        try:
            while True:
                msg = consumer.poll(1.0)
                if msg is not None and msg.error() is None:
                    try:
                        key = msg.key().decode("utf-8") if msg.key() else None
                        value = msg.value().decode("utf-8")
                        
                        # Try to parse as JSON
                        try:
                            parsed_value = json.loads(value)
                        except json.JSONDecodeError:
                            parsed_value = value
                        
                        message_data = {
                            "topic": msg.topic(),
                            "partition": msg.partition(),
                            "offset": msg.offset(),
                            "key": key,
                            "value": parsed_value,
                            "timestamp": msg.timestamp()
                        }
                        
                        yield message_data
                        
                    except Exception as e:
                        print(f"Error processing message: {e}")
                        
        except KeyboardInterrupt:
            print("Consumer interrupted by user")
        finally:
            consumer.close()

    def send_task_message(self, task_data: Dict[str, Any], topic: str = "automation-tasks"):
        """Helper method to send task messages to the automation topic"""
        return self.produce_message(topic, task_data, key=task_data.get('task_id'))

    def send_workflow_message(self, workflow_data: Dict[str, Any], topic: str = "workflow-events"):
        """Helper method to send workflow messages"""
        return self.produce_message(topic, workflow_data, key=workflow_data.get('workflow_id'))


# Example usage functions
def example_producer():
    """Example of how to use the Kafka client as a producer"""
    client = KafkaClient()
    
    sample_task = {
        "task_id": "task-001",
        "type": "data_processing",
        "status": "pending",
        "payload": {
            "input_file": "data.csv",
            "output_format": "json"
        },
        "created_at": "2024-01-01T10:00:00Z"
    }
    
    success = client.send_task_message(sample_task)
    if success:
        print("Task message sent successfully")
    else:
        print("Failed to send task message")

def example_consumer():
    """Example of how to use the Kafka client as a consumer"""
    client = KafkaClient()
    
    print("Starting to consume messages from automation-tasks topic...")
    for message in client.consume_messages("automation-tasks"):
        print(f"Received message: {message}")
        
        # Process the message based on its content
        if message['value'].get('type') == 'data_processing':
            print("Processing data processing task...")
        elif message['value'].get('type') == 'workflow_trigger':
            print("Processing workflow trigger...")

if __name__ == "__main__":
    import sys
    
    if len(sys.argv) > 1:
        if sys.argv[1] == "producer":
            example_producer()
        elif sys.argv[1] == "consumer":
            example_consumer()
        else:
            print("Usage: python kafka_client.py [producer|consumer]")
    else:
        print("EasyFlow Kafka Client")
        print("Usage:")
        print("  python kafka_client.py producer  - Run example producer")
        print("  python kafka_client.py consumer  - Run example consumer")

File: kafka-clients/nodejs/kafkaClient.js
const fs = require("fs");
const path = require("path");
const { Kafka } = require("@confluentinc/kafka-javascript").KafkaJS;

// ✅ INSTRUCTION 1: Import OpenTelemetry components for trace propagation
const { context, propagation, trace } = require('@opentelemetry/api');

class KafkaClient {
    constructor(configPath = "kafka-clients/nodejs/client.properties") {
        this.config = this.readConfig(configPath);
        this.producer = null;
        this.consumer = null;
        this.kafka = new Kafka();
        
        // Initialize W3C Trace Context Propagator
        this.propagator = propagation;
    }

    readConfig(fileName) {
        try {
            const data = fs.readFileSync(fileName, "utf8").toString().split("\n");
            return data.reduce((config, line) => {
                const [key, value] = line.split("=");
                if (key && value && !key.startsWith("#")) {
                    config[key.trim()] = value.trim();
                }
                return config;
            }, {});
        } catch (error) {
            console.warn(`Config file ${fileName} not found. Using environment variables.`);
            return {
                'bootstrap.servers': process.env.KAFKA_BOOTSTRAP_SERVERS || 'localhost:9092',
                'security.protocol': process.env.KAFKA_SECURITY_PROTOCOL || 'PLAINTEXT',
                'sasl.mechanisms': process.env.KAFKA_SASL_MECHANISMS || '',
                'sasl.username': process.env.KAFKA_SASL_USERNAME || '',
                'sasl.password': process.env.KAFKA_SASL_PASSWORD || '',
                'client.id': process.env.KAFKA_CLIENT_ID || 'easyflow-nodejs-client'
            };
        }
    }

    async initProducer() {
        if (!this.producer) {
            this.producer = this.kafka.producer(this.config);
            await this.producer.connect();
        }
        return this.producer;
    }

    async initConsumer(groupId = "easyflow-nodejs-group") {
        if (!this.consumer) {
            const consumerConfig = { ...this.config };
            consumerConfig["group.id"] = groupId;
            consumerConfig["auto.offset.reset"] = "earliest";
            
            this.consumer = this.kafka.consumer(consumerConfig);
            await this.consumer.connect();
        }
        return this.consumer;
    }

    /**
     * ✅ INSTRUCTION 1: Helper function to inject trace context into Kafka headers
     * Converts OpenTelemetry context to W3C traceparent/tracestate headers as Buffers
     */
    injectContextToKafkaHeaders(headers = {}) {
        const carrier = {};
        
        // Extract current active context and inject into carrier
        this.propagator.inject(context.active(), carrier);
        
        // Convert carrier to Kafka-compatible Buffer headers
        for (const [key, value] of Object.entries(carrier)) {
            if (value) {
                headers[key] = Buffer.from(String(value), 'utf-8');
            }
        }
        
        return headers;
    }

    /**
     * ✅ INSTRUCTION 2: Helper function to extract trace context from Kafka headers
     * Converts Kafka Buffer headers to OpenTelemetry context
     */
    extractContextFromKafkaHeaders(kafkaHeaders = {}) {
        const carrier = {};
        
        // Convert Kafka Buffer headers to string carrier
        for (const [key, value] of Object.entries(kafkaHeaders)) {
            if (value) {
                carrier[key] = Buffer.isBuffer(value) ? value.toString('utf-8') : String(value);
            }
        }
        
        // Extract context from carrier
        return this.propagator.extract(context.active(), carrier);
    }

    async produceMessage(topic, message, key = null) {
        try {
            const producer = await this.initProducer();
            
            const messageValue = typeof message === 'object' ? 
                JSON.stringify(message) : 
                String(message);

            // ✅ INSTRUCTION 1: Inject trace context into message headers
            const headers = this.injectContextToKafkaHeaders();

            const produceRecord = await producer.send({
                topic,
                messages: [{ 
                    key: key || `msg-${Date.now()}`, 
                    value: messageValue,
                    headers: headers  // Add trace headers to message
                }],
            });

            console.log(`Message produced to topic ${topic}:`, {
                key: key || `msg-${Date.now()}`,
                value: messageValue,
                partition: produceRecord[0].partition,
                offset: produceRecord[0].baseOffset,
                traceHeaders: Object.keys(headers).length > 0 ? 'injected' : 'none'
            });

            return produceRecord;
        } catch (error) {
            console.error("Error producing message:", error);
            throw error;
        }
    }

    async consumeMessages(topic, messageHandler, groupId = "easyflow-nodejs-group") {
        try {
            const consumer = await this.initConsumer(groupId);
            
            await consumer.subscribe({ topics: [topic] });

            // Setup graceful shutdown
            const disconnect = async () => {
                try {
                    await consumer.commitOffsets();
                    await consumer.disconnect();
                    if (this.producer) {
                        await this.producer.disconnect();
                    }
                } catch (error) {
                    console.error("Error during disconnect:", error);
                }
            };

            process.on("SIGTERM", disconnect);
            process.on("SIGINT", disconnect);

            await consumer.run({
                eachMessage: async ({ topic, partition, message }) => {
                    // ✅ INSTRUCTION 2: Extract trace context from incoming message
                    const extractedContext = this.extractContextFromKafkaHeaders(message.headers || {});
                    
                    // Create a new span with the extracted context as parent
                    const tracer = trace.getTracer('kafka-consumer');
                    
                    // Run message handler within the extracted trace context
                    await context.with(extractedContext, async () => {
                        const span = tracer.startSpan('kafka.consume', {
                            attributes: {
                                'messaging.system': 'kafka',
                                'messaging.destination': topic,
                                'messaging.operation': 'receive',
                                'messaging.kafka.partition': partition,
                                'messaging.kafka.offset': message.offset.toString(),
                                'messaging.message_id': message.key ? message.key.toString() : undefined
                            }
                        });

                        try {
                            const messageData = {
                                topic,
                                partition,
                                offset: message.offset,
                                key: message.key ? message.key.toString() : null,
                                value: message.value.toString(),
                                timestamp: message.timestamp,
                                headers: message.headers
                            };

                            // Try to parse JSON
                            try {
                                messageData.parsedValue = JSON.parse(messageData.value);
                            } catch {
                                messageData.parsedValue = messageData.value;
                            }

                            await messageHandler(messageData);
                            span.setStatus({ code: 1 }); // OK
                        } catch (error) {
                            span.recordException(error);
                            span.setStatus({ code: 2, message: error.message }); // ERROR
                            console.error("Error processing message:", error);
                        } finally {
                            span.end();
                        }
                    });
                },
            });
        } catch (error) {
            console.error("Error in consumer:", error);
            throw error;
        }
    }

    async sendTaskMessage(taskData, topic = "automation-tasks_") {
        const taskMessage = {
            taskId: taskData.taskId || `task-${Date.now()}`,
            type: taskData.type || "automation",
            status: taskData.status || "pending",
            payload: taskData.payload || {},
            createdAt: new Date().toISOString(),
            ...taskData
        };

        return this.produceMessage(topic, taskMessage, taskMessage.taskId);
    }

    async sendWorkflowMessage(workflowData, topic = "workflow-events") {
        const workflowMessage = {
            workflowId: workflowData.workflowId || `workflow-${Date.now()}`,
            event: workflowData.event || "trigger",
            status: workflowData.status || "pending",
            data: workflowData.data || {},
            timestamp: new Date().toISOString(),
            ...workflowData
        };

        return this.produceMessage(topic, workflowMessage, workflowMessage.workflowId);
    }

    async sendStatusUpdate(statusData, topic = "status-updates") {
        const statusMessage = {
            entityId: statusData.entityId,
            entityType: statusData.entityType || "task",
            status: statusData.status,
            message: statusData.message || "",
            timestamp: new Date().toISOString(),
            ...statusData
        };

        return this.produceMessage(topic, statusMessage, statusMessage.entityId);
    }

    async disconnect() {
        try {
            if (this.consumer) {
                await this.consumer.commitOffsets();
                await this.consumer.disconnect();
                this.consumer = null;
            }
            if (this.producer) {
                await this.producer.disconnect();
                this.producer = null;
            }
        } catch (error) {
            console.error("Error during disconnect:", error);
        }
    }
}

// Example usage functions
async function exampleProducer() {
    const client = new KafkaClient();
    
    try {
        const sampleTask = {
            taskId: "task-node-001",
            type: "automation_run",
            status: "pending",
            payload: {
                script: "data_processing.js",
                parameters: { inputFile: "data.csv" }
            }
        };

        await client.sendTaskMessage(sampleTask);
        console.log("Task message sent successfully");
        
        const workflowEvent = {
            workflowId: "workflow-001",
            event: "started",
            status: "running",
            data: { initiatedBy: "user123" }
        };

        await client.sendWorkflowMessage(workflowEvent);
        console.log("Workflow message sent successfully");
        
    } catch (error) {
        console.error("Producer example failed:", error);
    } finally {
        await client.disconnect();
    }
}

async function exampleConsumer() {
    const client = new KafkaClient();
    
    try {
        console.log("Starting to consume messages from automation-tasks topic...");
        
        await client.consumeMessages("automation-tasks_", async (message) => {
            console.log("Received task message:", {
                key: message.key,
                value: message.parsedValue,
                offset: message.offset,
                partition: message.partition
            });
            
            // Process based on message type
            if (message.parsedValue.type === 'automation_run') {
                console.log("Processing automation run task...");
                // Add your automation logic here
            } else if (message.parsedValue.type === 'data_processing') {
                console.log("Processing data processing task...");
                // Add your data processing logic here
            }
        });
        
    } catch (error) {
        console.error("Consumer example failed:", error);
    }
}

// CLI interface
if (require.main === module) {
    const command = process.argv[2];
    
    if (command === "producer") {
        exampleProducer();
    } else if (command === "consumer") {
        exampleConsumer();
    } else {
        console.log("EasyFlow Kafka Node.js Client");
        console.log("Usage:");
        console.log("  node kafkaClient.js producer  - Run example producer");
        console.log("  node kafkaClient.js consumer  - Run example consumer");
    }
}

module.exports = KafkaClient;